designs/2023-rule-performance-dashboard/README.md

```diff
@@ -0,0 +1,210 @@
+
+  ```bash
+  TIMING=1 DEBUG=eslint:cli-engine eslint --no-eslintrc --rule ... file
+  ```
+
+2. Doing a single ESLint run on all files and then extracting the file/rule time data output:
+
+  ```bash
+  TIMING=all DEBUG=eslint:cli-engine eslint ... files
+  ```
+
+In addition, one needs to create an overview for an effective presentation of the results.
+
+It would be more *convenient and shareable* to have:
+
+   1. The **timing data** ESLint already collects exposed to the formatters
+   2. A **built-in formatter** for that data
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+The *proof-of-concept* can be found at:
+- [**ESLint**](https://github.com/mnkiefer/eslint/pull/1): **Fork of ESLint** with the above mentioned changes.
+- [**Sample project**](https://github.com/mnkiefer/eslint-samples): Example on which the dashboard (shown below) was generated.
+
+### ESLint *timing* exposed
+
+Exposure of the timing object requires only a few changes in the Linter and CLI engline.
+
+The timing is already collected in `lib/linter/timing.js`. Note, that the function requires the extra input paramter `filename`. For the function itself, the only change is that we now *persist* that information in the `lintTimesPerRule` object, which stores the detailed *per file per rule* lint times under the **lintTimes** key. An extra function `getLintTime()` is returned to the linter to retrieve this object.
+
+The linter `lib/linter/linter.js` retrieves the timing data only when `TIMING` is enabled (i.e. from the ESLint CLI).
+
+Finally, we append the new [timing data](#timing-data) to the *Lint Report* in the CLI engine `lib/cli-engine/cli-engine.js` such that they are accessible to the formatters:
+- `lintTime` (ms): Lint time of file
+- `lintOrder` (order number): Lint order of file
+- `lintTimePerRule` (ms): Lint time of file per rule
+
+Below is an excerpt of a sample for the 5th file that was linted in the sample project:
+
+```json
+    {
+        ...
+        "lintOrder": 5,
+        "lintTime": 13.642954999999999,
+        "lintTimePerRule": {
+            "constructor-super": 0.05145500000000001,
+            "for-direction": 0.032875,
+            "getter-return": 0.003292,
+            "no-async-promise-executor": 0,
+            "no-case-declarations": 0,
+            "no-class-assign": 0,
+            "no-compare-neg-zero": 0,
+            "no-cond-assign": 0.035042000000000004,
+            "no-const-assign": 0.0025,
+            "no-constant-condition": 0.563125,
+            ...
+        },
+        ...
+        "fileSize": 312
+    }
+```
+
+### Formatter `html-charts`
```

Comment:
It seems like this should be something more specific, like `html-rule-performance`?

Reply:
I agree. Will update the text accordingly. 👍🏼